json.lua

```diff
@@ -64,32 +64,43 @@ local function encode_table(val, stack)
   if stack[val] then error("circular reference") end
 
   stack[val] = true
+  -- Check whether to treat as a array or object
+  local array = true
+  local length = 0
+	local nLen = 0
+  for k,v in pairs(val) do
+		if (type(k) ~= "number" or k<=0) and not (k == "n" and type(v) == "number") then
+			array = nil
+			break	-- Treat as object
+		else
+			if k > length then 
+				length = k
+			end
+			if k == "n" and type(v) == "number" then
+				nLen = v
+			end
+		end
+  end
 
-  if rawget(val, 1) ~= nil or next(val) == nil then
-    -- Treat as array -- check keys are valid and it is not sparse
-    local n = 0
-    for k in pairs(val) do
-      if type(k) ~= "number" then
-        error("invalid table: mixed or invalid key types")
-      end
-      n = n + 1
-    end
-    if n ~= #val then
-      error("invalid table: sparse array")
-    end
+  if array then
+		if nLen > length then
+			length = nLen
+		end
     -- Encode
-    for i, v in ipairs(val) do
-      table.insert(res, encode(v, stack))
+    for i=1,length do
+      table.insert(res, encode(val[i], stack))
     end
     stack[val] = nil
     return "[" .. table.concat(res, ",") .. "]"
 
   else
     -- Treat as an object
     for k, v in pairs(val) do
+	--[[
```

Comment:
Dirty

Reply:
Don't understand. Perhaps a more useful/productive comment?

Reply:
Ok, thanks. I usually keep some portions in my code to see some history immediately but you are right it is in the version history. I cleaned it up.